File: boilerplate/python/order-service/src/application/usecases/get_order_use_case.py
"""
Get Order Use Case demonstrating Cache-Aside pattern.

Pattern:
1. Check cache first
2. On cache miss → load from database
3. Populate cache with result
4. Return order

This optimizes read-heavy workloads by reducing database queries.
"""

from typing import Optional
from domain.models.order import Order
from domain.ports.order_repository import OrderRepository
from domain.ports.cache_manager import CacheManager
from domain.order_id import OrderId


class GetOrderUseCase:
    """Get order by ID with cache-aside pattern."""
    
    CACHE_KEY_PREFIX = "order-service:order:"
    CACHE_KEY_SUFFIX = ":full"
    
    def __init__(self, order_repository: OrderRepository, cache_manager: CacheManager):
        """
        Initialize use case.
        
        Args:
            order_repository: Order repository port
            cache_manager: Cache manager port
        """
        self.order_repository = order_repository
        self.cache_manager = cache_manager
    
    def execute(self, order_id: OrderId) -> Order:
        """
        Get order by ID with cache-aside pattern.
        
        Args:
            order_id: Order ID
            
        Returns:
            Order if found
            
        Raises:
            OrderNotFoundException: If order doesn't exist
        """
        cache_key = self._build_cache_key(order_id)
        
        # Step 1: Check cache first
        cached_order = self.cache_manager.get(cache_key, Order)
        if cached_order is not None:
            return cached_order
        
        # Step 2: Cache miss - load from database
        order = self.order_repository.find_by_id(order_id)
        if order is None:
            raise OrderNotFoundException(order_id)
        
        # Step 3: Populate cache
        self.cache_manager.put(cache_key, order)
        
        # Step 4: Return order
        return order
    
    def invalidate_cache(self, order_id: OrderId) -> None:
        """
        Invalidate order cache (call after updates).
        
        Args:
            order_id: Order ID to invalidate
        """
        cache_key = self._build_cache_key(order_id)
        self.cache_manager.evict(cache_key)
    
    def _build_cache_key(self, order_id: OrderId) -> str:
        """Build cache key from order ID."""
        return f"{self.CACHE_KEY_PREFIX}{order_id.value}{self.CACHE_KEY_SUFFIX}"


class OrderNotFoundException(Exception):
    """Raised when order is not found."""
    
    def __init__(self, order_id: OrderId):
        super().__init__(f"Order not found: {order_id.value}")

File: boilerplate/java/order-service/src/main/java/com/example/orderservice/application/usecases/GetOrderUseCase.java
package com.example.orderservice.application.usecases;

import com.example.orderservice.domain.models.Order;
import com.example.orderservice.domain.ports.OrderRepository;
import com.example.orderservice.domain.ports.CacheManager;
import com.example.orderservice.domain.models.OrderId;
import org.springframework.stereotype.Component;

import java.util.Optional;

/**
 * Get Order Use Case demonstrating Cache-Aside pattern.
 * 
 * Pattern:
 * 1. Check cache first
 * 2. On cache miss → load from database
 * 3. Populate cache with result
 * 4. Return order
 * 
 * This optimizes read-heavy workloads by reducing database queries.
 */
@Component
public class GetOrderUseCase {
    
    private static final String CACHE_KEY_PREFIX = "order-service:order:";
    private static final String CACHE_KEY_SUFFIX = ":full";
    
    private final OrderRepository orderRepository;
    private final CacheManager cacheManager;
    
    public GetOrderUseCase(OrderRepository orderRepository, CacheManager cacheManager) {
        this.orderRepository = orderRepository;
        this.cacheManager = cacheManager;
    }
    
    /**
     * Get order by ID with cache-aside pattern.
     * 
     * @param orderId Order ID
     * @return Order if found
     * @throws OrderNotFoundException if order doesn't exist
     */
    public Order execute(OrderId orderId) {
        String cacheKey = buildCacheKey(orderId);
        
        // Step 1: Check cache first
        Optional<Order> cachedOrder = cacheManager.get(cacheKey, Order.class);
        if (cachedOrder.isPresent()) {
            return cachedOrder.get();
        }
        
        // Step 2: Cache miss - load from database
        Order order = orderRepository.findById(orderId)
            .orElseThrow(() -> new OrderNotFoundException(orderId));
        
        // Step 3: Populate cache
        cacheManager.put(cacheKey, order);
        
        // Step 4: Return order
        return order;
    }
    
    /**
     * Invalidate order cache (call after updates).
     * 
     * @param orderId Order ID to invalidate
     */
    public void invalidateCache(OrderId orderId) {
        String cacheKey = buildCacheKey(orderId);
        cacheManager.evict(cacheKey);
    }
    
    private String buildCacheKey(OrderId orderId) {
        return CACHE_KEY_PREFIX + orderId.getValue() + CACHE_KEY_SUFFIX;
    }
    
    public static class OrderNotFoundException extends RuntimeException {
        public OrderNotFoundException(OrderId orderId) {
            super("Order not found: " + orderId.getValue());
        }
    }
}
